docs/transaction.md

```diff
@@ -0,0 +1,261 @@
+
+However, "RENAME" requires two micro-ops: fetching the data from two keys, and then the second hop - deleting/writing a key (depending whether the key is a source or a destination).
+
+When a coordinator schedules an execution, it also specifies whether this execution is the last hop for that command. This is necessary, so that shards could do clean-up operations when running the last execution request: unlocking the keys and removing the transaction from the tx-queue.
+
+Once a coordinator sends the micro-op request to all the shards, it waits for an answer. Only when all shards executed the micro-op and return the result, the coordinator is unblocked and it can proceed to the next hop. The coordinator is allowed to process the intermediary responses from the previous hops in order to define the next execution request.
+
+The shards always execute transactions at the head of the tx-queue. When the last execution hop for that transaction is executed the transaction is removed from the queue and the next one can be executed. This way we maintain the ordering guarantees specified by the scheduling order of the transactions and we maintain the atomicity of operations across multiple shards.
+
+## Multi-op transactions (Redis transactions)
+
+Redis transactions (MULTI/EXEC sequences) are modelled as `N` commands within a Dragonfly transaction. In order to avoid ambiguity with terms, we call a Redis transaction -  multi-transaction in Dragonfly.
+
+Dragonfly transactional framework allows running any command as standalone transaction or within a multi-transaction. The complexity of handling both cases is pushed into the transactional framework and a command implementation is not aware of a transaction context. Take for example, redis transaction consisting of 2 commands:
+
+```
+MULTI
+SET x foo
+SET y bar
+EXEC
+```
+
+`SET` command can be implemented with two hops (below is a pseudo-code):
+
+```cpp
+trans->Schedule();   // Hop1
+
+trans->Exec(trans, [](key, val) {
+  dict[key] = val;
+});   // Hop2
+```
+
+The commands implementation layer does not differentiate between muti-tx and regular txs.
+Therefore, we have the following hard-coded flow:
+
+```cpp
+// set x foo
+trans->Schedule();
+
+trans->Exec([](key, val) {
+  dict[key] = val;    // x <- foo
+});
+
+// set y bar
+trans->Schedule();     // noop for multi-tx.
+trans->Exec([](key, val) {
+  dict[key] = val;    // y <- bar
+});
+```
+
+However, we need to keep the same tx for both `SET` operations and schedule it only once.
+The transactional framework maintains an internal state underneath such that it avoids rescheduling
+a multi-tx for the second command "set y bar" and makes the second call to `Schedule()` effectively a noop.
+
+## Optimizations
+Out of order transactions - TBD
+
+## Blocking commands (BLPOP)
+
+Redis has a rich api with around 200 commands. Few of those commands provide blocking semantics, which allow using Redis as publisher/subscriber broker.
```

Comment:
Sorry this is to much reading for me to digest in one day.. but
Is this section related to how DragonflyDB transactions framework provide serializability for its multi-key and multi-command operations?

Reply:
no, I will move it to another document in subsequent Prs.